crates/wasmtime/src/component/values.rs

```diff
@@ -0,0 +1,582 @@
+        }
+
+        Ok(Self {
+            ty: ty.clone(),
+            values,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Variant {
+    pub(crate) ty: types::Variant,
+    pub(crate) discriminant: u32,
+    pub(crate) value: Box<Val>,
+}
+
+impl Variant {
+    /// Instantiate the specified type with the specified case `name` and `value`.
+    pub fn try_new(ty: &types::Variant, name: &str, value: Val) -> Result<Self> {
+        let (discriminant, case_type) = ty
+            .cases()
+            .enumerate()
+            .find_map(|(index, case)| {
+                if case.name == name {
+                    Some((index, case.ty))
+                } else {
+                    None
+                }
+            })
+            .ok_or_else(|| anyhow!("unknown variant case: {name}"))?;
+
+        case_type
+            .check(&value)
+            .with_context(|| format!("type mismatch for case {name} of variant"))?;
+
+        Ok(Self {
+            ty: ty.clone(),
+            discriminant: u32::try_from(discriminant)?,
+            value: Box::new(value),
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Enum {
+    pub(crate) ty: types::Enum,
+    pub(crate) discriminant: u32,
+}
+
+impl Enum {
+    /// Instantiate the specified type with the specified case `name`.
+    pub fn try_new(ty: &types::Enum, name: &str) -> Result<Self> {
+        let discriminant = u32::try_from(
+            ty.names()
+                .position(|n| n == name)
+                .ok_or_else(|| anyhow!("unknown enum case: {name}"))?,
+        )?;
+
+        Ok(Self {
+            ty: ty.clone(),
+            discriminant,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Union {
+    pub(crate) ty: types::Union,
+    pub(crate) discriminant: u32,
+    pub(crate) value: Box<Val>,
+}
+
+impl Union {
+    /// Instantiate the specified type with the specified `discriminant` and `value`.
+    pub fn try_new(ty: &types::Union, discriminant: u32, value: Val) -> Result<Self> {
+        if let Some(case_ty) = ty.types().nth(usize::try_from(discriminant)?) {
+            case_ty
+                .check(&value)
+                .with_context(|| format!("type mismatch for case {discriminant} of union"))?;
+
+            Ok(Self {
+                ty: ty.clone(),
+                discriminant,
+                value: Box::new(value),
+            })
+        } else {
+            Err(anyhow!(
+                "discriminant {discriminant} out of range: [0,{})",
+                ty.types().len()
+            ))
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Option {
+    pub(crate) ty: types::Option,
+    pub(crate) discriminant: u32,
+    pub(crate) value: Box<Val>,
+}
+
+impl Option {
+    /// Instantiate the specified type with the specified `value`.
+    pub fn try_new(ty: &types::Option, value: std::option::Option<Val>) -> Result<Self> {
```

Comment:
Idiomatically I think it's fine to call these constructors `new` rather than `try_new` since `try_*` is typically only used if the base name is already a method elsewhere (and otherwise there's no `Option::new`)